### ISSUE TITLE
More transfer variables

### ISSUE THREAD

**Issue by DeputyLOL** [Mihara/RasterPropMonitor#452]
Currently we can transfer persistents to a plugin. But not customs, mapped, math or select.
There are some occasions where you need to change a persistent variable and then transfer it, preferably in one button click.

**Comment by MOARdV** [comment]
Yeah, that was on my "I should fix that" mental list.  Now it's on the issues list, so I'll be less likely to forget it. :)

**Comment by DeputyLOL** [comment]
Do you think we could have a 'transfer on change' method. Meaning it would transfer automatically whenever the input variable changes.
Please let me know.

**Comment by MOARdV** [comment]
> Do you think we could have a 'transfer on change' method. Meaning it would transfer automatically whenever the input variable changes.

It's possible, but it would need to be a different thing, not part of the action group switch module.

**Comment by DeputyLOL** [comment]
ok thank you

**Comment by DeputyLOL** [comment]
Do you think it would be possible to make transfers like triggers?
So that they are controlled by variables rather than buttons, it would help a great deal as the current method is very limited and unreliable. It would also allow more than one transfer at once.

**Comment by MOARdV** [comment]
I don't think automatic variable transfers would be a good idea - at least, not to replace current behavior.  It might be possible to add something that automatically does the transfer every update, but it'd require some thought on how to implement it.

**Comment by DeputyLOL** [comment]
I find that using the transfer variables on buttons irritating as they activate when the button is pressed AND released, the block transfer feature works only occasionally and the transfer only accepts persistents and is only able to be activated by pressing a button. 
The current method we have works for simple things, but it can be very challenging when your trying to make more complex systems. 

Having the transfers as a separate module would be an enormous help as it could allow you to use custom variables to trigger it, meaning it will trigger only when that variable is on.

**Comment by DeputyLOL** [comment]
I'm also curious about why automatic transfers wouldn't be a good idea? Is it a lot of work or resource hungry? I mainly want it as a separate module so I have more control over when it triggers and can have more than one trigger at the same time as buttons don't allow that. 
As for the update on change thing, I figured out a way to do that on my numerical inputs. Simply set the colliders as an monetary contact switch and put a transfer on both. :D

**Comment by MOARdV** [comment]
The main issue with automatic transfers is that it would be processed every physics update - so, 10Hz typically.  The variable would have to be queried, and if its value changed, it'd need to be sent to the plugin (or wherever the change needs to go).  That gets hairier when dealing with floating point values, since a lot of variables in KSP are noisy - there are small changes every update, such as Ap and Pe in orbit, or ASL in flight.  If there is any difference between the current value and the previous value (like a 1mm change in ASL), RPM would update the plugin, which may lead to side effects (for instance, Pilot Assistant sometimes resets its autopilots if a value gets changed; it may also be computationally costly if the plugin has to recompute other values based on the updated results).

If there's a controlling variable that decides if the variable is to be updated, then there's the additional queries that need to take place (3 more - current value of the controlling variable, plus the two ends of the range).  At least as far as I've used transfer variables, I don't see a need to continually update their value - the ones I've used have been "set once" type variables.  If you've got a use case that shows why continually-updated transfer variables are needed, please let me know.

In summary: no, it's not out of the question, but, yes, I will need to sort out the changes required to support it.  Since it's not a feature I need right now, it's not a priority until I have time to devote to KSP again (Real Life has been very busy, and will be for the foreseeable future).

**Comment by DeputyLOL** [comment]
Just for clarity, I'm not talking about automatically updating transfers anymore. :/
I found a way to do that already.

I'm talking about having a transfer module that only transfers when a certain variable turns from 0 to 1
Just like the trigger module works. 
It only needs the name of the handler (JSIPilotAssistant), the source variable (getvariable)or method (getmethod), the target method(setmethod) or variable (setvariable) and the trigger variable + range. :D

I'm trying to make the autopilot in my cockpit and there are a few problems that can be fixed by not having transfers attached to buttons as they transfer when the button is activated (pressed) and deactivated (pressed again) which causes a lot of problems.

Let me know what you think.

**Comment by DeputyLOL** [comment]
I am best at explaining things by writing down what the config could look like.
Could look something like this:

RPM_TRANSFER_EVENT
{
   name = JSIPilotAssistant
   variableName = CUSTOM_TRANSFER
   range = 0.5,1.5
   getVariable(getMethod) = PERSISTENT_ALT/MATH_ALT_AMOUNT/SELECT_ALT_MODE
   setMethod(setVariable) = SetVertAltitude
}

Very similar to trigger config.

**Comment by MOARdV** [comment]
> I'm trying to make the autopilot in my cockpit and there are a few problems that can be fixed by not having transfers attached to buttons as they transfer when the button is activated (pressed) and deactivated (pressed again) which causes a lot of problems.

There shouldn't be two transfer events.  There should only be one.  If you're seeing a double transfer, please post a config for that prop and I'll see if I can figure out what's happening.  Press/release behavior on buttons only works if the button is configured to behave that way.  (Actually, if there are multiple props, such as for the pilot and copilot, it could trigger more than once if the controls are tied together, since triggering one button would trigger the other one).

As far as your proposed design, it's certainly possible.  Is the idea that the transfer happens always when CUSTOM_TRANSFER is in range, or only when CUSTOM_TRANSFER transitions from 'off' to 'on'?

**Comment by DeputyLOL** [comment]
I think it would be better to have to it transfer when the custom transitions from off to on. like the triggers do.

**Comment by DeputyLOL** [comment]
As for the buttons, I got them to work but its not always reliable. Ill try and explain it.
Basically I wanted to be able to buttons that control autopilot modes. Only one can be turned on at a time, but all can be turned off.
There are 2 buttons on the autopilot. ALT HOLD and V/S (you can probably guess what they mean).
When I press the ALT HOLD button, the autopilot switches to Altitude mode and transfers the number I have set it to hold. 
If I press the button again, it stays in altitude mode but the vertical autopilot turns off. To change the altitude I have a separate button called LVL CHG (not important).
If I want to hold a vertical speed I press V/S. To change it I simple change the set V/S and it is automatically transferred with each click.  Pressing it again turns off the vertical autopilot.
The real problem was that I wanted to be able to switch modes with one button press. So instead of having to turn one mode off then the other on, I would do it for me. When I press the button to switch modes the following would happen;
1. V/S Button pushed -> transfers to autopilot switching its mode to vertspeed.
2. The ALT HOLD button is turned off as both buttons are on and the autopilot isn't in Altitude mode.
3. As the ALT HOLD button is turned off it transfers its setting to the autopilot switching it back to altitude mode.
4. The autopilot is now still in alt mode with the V/S button down, if I press the ALT HOLD button, then it will switch to V/S. Only fix is to turn off both and then turn one on.

Occasionally it works as intended and is very responsive when it does.
This is the ALT HOLD button config, VS is almost Identical but with different variables.

	MODULE
	{
		name = JSIActionGroupSwitch
		animationName = switch_B_ButtonPushAnim
		switchTransform = switch_B_ButtonBody
		actionName = transfer
		masterVariableName = CUSTOM_AUTOP_ALTHOLD_BUTTON_MASTER
		masterVariableRange = -0.5, 0.5
		//momentarySwitch = true
		//switchGroupIdentifier = 3
		perPodPersistenceName = AUTOP_ALTHOLD_BUTTON //AUTOP_VERTMODE_BUTTON
		switchSound = AdvancedCockpit/Sounds/B_Button01_Click
		TRANSFERACTION
		{
		    name = JSIPilotAssistant
		    setMethod = SetVertAltitude
		    perPodPersistenceName = HOLDALT
		    stateVariable = CUSTOM_AUTOP_ALTHOLD_BUTTON_TRANSFER
		}
	}

Even though this method 'works' now, I would prefer the transfers to be separate from the buttons to make the whole system much more simple.

**Comment by DeputyLOL** [comment]
"As far as your proposed design, it's certainly possible. Is the idea that the transfer happens always when CUSTOM_TRANSFER is in range, or only when CUSTOM_TRANSFER transitions from 'off' to 'on'?"

Thinking about the proposed design, I an option for both 'transfer once' and 'transfer while on' would be appreciated as I can think of some really nice features that could be done with it.
I know that your busy with stuff and KSP 1.1 and that this isn't a top priority at the moment. Just wanted to write this down here before I forget it completely. :)

**Comment by MOARdV** [comment]
The original request (transfer variables directly from other variables) will be in v0.25.0.  The extended features will need to wait - I think the JSISwitch function will most likely be able to meet most of those requests when I get it implemented.

**Comment by DeputyLOL** [comment]
i had an idea about the switches, I don't know how hard it would be to implement.
Make the switch action its own module that can be toggle by variable. This would include the plugins, stock functions(gear,sas...) and transfers, all switches would then become dummys. The actions would then be its own module and allow any variable to toggle them. It would probably function similar to the trigger variables but would end up replacing them.

**Comment by DeputyLOL** [comment]
Also as an answer to a previous comment:
"If you've got a use case that shows why continually-updated transfer variables are needed, please let me know."
Now that Nav Utilities' variables are usable in rpm, I might be able to put together an autoland or approach system which would be fed into pilot assistant. 
I don't exactly know how pilot assistant would deal with the transfers as its behaviour is questionable sometimes.

**Comment by MOARdV** [comment]
> i had an idea about the switches, I don't know how hard it would be to implement. Make the switch action its own module that can be toggle by variable

That's a pretty good idea.  Instead of adding another switch module, I can add a new action to the existing switches.  I still will have to add a module that can handle automatic actions, although maybe that can be added to an existing module - RasterPropMonitorComputer is a good candidate, since it's withered to very little functionality now (it's not really required for anything, other than vessel descriptions, stored strings, triggered events, and enabling debug logging).  I'll see what I can come up with.

**Comment by DeputyLOL** [comment]
Nice :D

**Comment by MOARdV** [comment]
The compound actions will be implemented as part of revamped issue #439.
